components/speedreader/speedreader_rewriter_service.cc

```diff
@@ -57,36 +63,77 @@ std::string GetDistilledPageStylesheet(const base::FilePath& stylesheet_path) {
+SpeedreaderRewriterService::SpeedreaderRewriterService()
+    : speedreader_(new speedreader::SpeedReader) {
   // Load the built-in stylesheet as the default
   content_stylesheet_ = WrapStylesheetWithCSP(
       ui::ResourceBundle::GetSharedInstance().LoadDataResourceString(
           IDR_SPEEDREADER_STYLE_DESKTOP));
 
-  // Check the paths from the component as observer may register
-  // later than the paths were available in the component.
-  const auto stylesheet_path = component_->GetStylesheetPath();
-  if (!stylesheet_path.empty())
-    OnStylesheetReady(stylesheet_path);
-
-  component_->AddObserver(this);
+  if (base::CommandLine::ForCurrentProcess()->HasSwitch(
+          kSpeedreaderStylesheet)) {
+    stylesheet_override_path_ =
+        base::CommandLine::ForCurrentProcess()->GetSwitchValuePath(
+            kSpeedreaderStylesheet);
+    watch_task_runner_ =
+        base::ThreadPool::CreateSequencedTaskRunner({base::MayBlock()});
+
+    // Manually trigger the initial stylesheet load
+    OnFileChanged(stylesheet_override_path_, false);
+
+    base::PostTaskAndReplyWithResult(
+        watch_task_runner_.get(), FROM_HERE,
+        base::BindOnce(
+            &CreateAndStartFilePathWatcher, stylesheet_override_path_,
+            base::BindPostTask(
+                base::SequencedTaskRunnerHandle::Get(),
+                base::BindRepeating(&SpeedreaderRewriterService::OnFileChanged,
+                                    weak_factory_.GetWeakPtr()))),
+        base::BindOnce(&SpeedreaderRewriterService::OnWatcherStarted,
+                       weak_factory_.GetWeakPtr()));
+  }
 }
 
 SpeedreaderRewriterService::~SpeedreaderRewriterService() {
-  component_->RemoveObserver(this);
+  if (file_watcher_) {
+    watch_task_runner_->DeleteSoon(FROM_HERE, file_watcher_);
+  }
+}
+
+void SpeedreaderRewriterService::OnFileChanged(const base::FilePath& path,
+                                               bool error) {
+  DCHECK_EQ(path, stylesheet_override_path_);
+  if (!error) {
+    base::ThreadPool::PostTaskAndReplyWithResult(
+        FROM_HERE, {base::MayBlock()},
+        base::BindOnce(&GetDistilledPageStylesheet, path),
+        base::BindOnce(&SpeedreaderRewriterService::OnLoadStylesheet,
+                       weak_factory_.GetWeakPtr()));
+  }
 }
 
-void SpeedreaderRewriterService::OnStylesheetReady(const base::FilePath& path) {
-  base::ThreadPool::PostTaskAndReplyWithResult(
-      FROM_HERE, {base::MayBlock()},
-      base::BindOnce(&GetDistilledPageStylesheet, path),
-      base::BindOnce(&SpeedreaderRewriterService::OnLoadStylesheet,
-                     weak_factory_.GetWeakPtr()));
+void SpeedreaderRewriterService::OnWatcherStarted(
+    base::FilePathWatcher* file_watcher) {
+  file_watcher_ = file_watcher;
+}
+
+void SpeedreaderRewriterService::OnLoadStylesheet(std::string stylesheet) {
+  VLOG(2) << "Speedreader stylesheet loaded";
+  content_stylesheet_ = stylesheet;
```

Comment:
Should be wrapped with WrapStylesheetWithCSP

Reply:
Should already be since it's called from `OnFileChanged`, which passes the result from `GetDistilledPageStylesheet`

Reply:
Yes, I missed it.

Reply:
Yes, I missed it.